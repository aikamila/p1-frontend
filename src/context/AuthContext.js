import { createContext, useState, useEffect, useCallback } from "react";
import { useHistory, useLocation } from "react-router-dom";
import { getItemAndCheckExpiry, setItemwith8MonthExpiry } from "../utils/LocalStorage";
import { calculateTime } from "../utils/Timing";
import { decodeUser } from "../utils/JWTDecode";

const AuthContext = createContext()

export default AuthContext;

export const AuthProvider = ({children}) => {

    const history = useHistory()

    let location = useLocation()

    const [authTokens, setAuthTokens] = useState(null)
    const [userId, setUserId] = useState(null)

    const [loginEmail, setLoginEmail] = useState('')
    const [loginPassword, setLoginPassword] = useState('')

    // const [pwResetEmail, setPwResetEmail] = useState('')
    // const [pwResetPassword, setPwResetPassword] = useState('')
    // const [pwResetPasswordSecond, setPwResetPasswordSecond] = useState('')

    const [loginError, setLoginError] = useState(null) // invalid credentials
 
    const [authServerError, setAuthServerError] = useState(null) // server error while logging in
    const [authSignUpServerError, setAuthSignUpServerError] = useState(null) // server error while signing up
    const [unexpectedLogoutError, setUnexpectedLogoutError] = useState('Something went wrong and you were logged out.')
    const [initialVerificationError, setInitialVerificationError] = useState(null)
    const [initialVerificationServerError, setInitialVerificationServerError] = useState(null)
    const [initialVerificationSuccessful, setInitialVerificationSuccessful] = useState(true)

    const [loginButtonDisabled, setLoginButtonDisabled] = useState(false)

    const BACKEND_DOMAIN = 'https://arcane-spire-03245.herokuapp.com/'
    const TOKENS_OBTAIN_URL = BACKEND_DOMAIN + 'api/user/token/'
    const TOKENS_REFRESH_URL = BACKEND_DOMAIN + 'api/user/token/refresh/'
    const TOKENS_BLACKLIST_URL = BACKEND_DOMAIN + 'api/user/token/blacklist/'
    const INITIAL_EMAIL_VERIFICATION = BACKEND_DOMAIN + 'api/user/token/verification/'

    const login = useCallback(async (e) => {
        e.preventDefault();
        setLoginButtonDisabled(true)
        try{
            const response = await fetch(TOKENS_OBTAIN_URL ,{
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({'email': loginEmail, 'password': loginPassword})
            })
            const data = await response.json()
            if (response.status === 401) {
                throw Error('Invalid credentials')
            }else if (response.status === 200){
                setUserId(decodeUser(data))
                setItemwith8MonthExpiry('authTokens', data)
                setAuthTokens(data)
                setLoginPassword('')
                history.push('/home')
            }else if(loginEmail.length === 0 || loginPassword.length === 0){
                throw Error('Invalid credentials')
            }else{
                throw Error('Something else happened')
            }
        }catch (err){
            switch (err.message){
                case 'Invalid credentials':
                    setLoginError(err.message)
                    setLoginPassword('')
                    break;
                default:
                    setAuthServerError('Upps... unexpected problem with the server occured. Try to refresh the site and log in again :)')
            }
        }
        setLoginButtonDisabled(false)
    }, 
    [TOKENS_OBTAIN_URL, loginEmail, loginPassword, history]
    )

    const logout = useCallback(async () => {
        try{
            await fetch(TOKENS_BLACKLIST_URL,{
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({'refresh': authTokens.refresh})
            })}catch(err){
                // doing nothing if the blacklisting of tokens goes wrong on the backend
            }
        const tokens = localStorage.getItem('authTokens')
        if (tokens){
            localStorage.removeItem('authTokens')
        }
        setAuthTokens(null)
        setUserId(null)
        history.push('/')
    }, [TOKENS_BLACKLIST_URL])

    const updateToken = useCallback(async () => {
        try{
            const tokens = getItemAndCheckExpiry('authTokens')
            // we get them from the local storage bc when the user comes back, they don't have the hook set (so we would send null :( )
            const response = await fetch(TOKENS_REFRESH_URL ,{
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({'refresh': tokens.refresh})
            })
            const data = await response.json()
            if (response.status !== 200) {
                throw Error('Invalid credentials')
            }else{
                setItemwith8MonthExpiry('authTokens', data)
                setAuthTokens(data)
                setUserId(decodeUser(data))
            }
        }catch(err){
            logout(); 
            setUnexpectedLogoutError('Something went wrong and you were logged out.')
            setTimeout(() => {
                setUnexpectedLogoutError(null)
            }, 5000)
        }
    }, [TOKENS_REFRESH_URL, logout]
    )

    const verifyEmail = async (id, token) => {
        try{
            const response = await fetch(INITIAL_EMAIL_VERIFICATION, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({'id': id, 'token': token})
                })
            if(response.status === 401){
                throw Error('Invalid link')
            }else if(response.status !== 200){
                throw Error("Something else happened")
            }
            const data = await response.json()
            if(response.status===200){
                await logout(); // if there is a user that is logged in, their token should be blacklisted
                setAuthTokens(data);
                setUserId(decodeUser(data))
                setItemwith8MonthExpiry('authTokens', data)
                setInitialVerificationSuccessful(true)
                setTimeout(() => {
                    setInitialVerificationSuccessful(false)
                }, 5000)
                history.push('/home')
            }
            }catch(err){
                if(err.message === 'Invalid link'){
                    setInitialVerificationError('The link is invalid. If you have already used it, just log into your account.')
                }else(
                    setInitialVerificationServerError('Something went wrong. Please use the link once again.')
                )
                history.push("/verification/unsuccessful")
            }
        
    }

    useEffect(() => {
        const fourMinutes = calculateTime(0, 4, 0)
        let interval = setInterval(() => {
            if (authTokens){
                updateToken()
            }
        }, fourMinutes)
        return () => clearInterval(interval)
    }, [authTokens, updateToken])

    useEffect(() => {
        const updateTokenAtTheBeginning = async () => {
            const tokens = getItemAndCheckExpiry('authTokens')
            if(tokens !== null){
                await updateToken();
            }
        }   
        updateTokenAtTheBeginning()
    }, [updateToken])

    useEffect(() => {
        setLoginPassword('')
        setLoginError(null)
        setAuthSignUpServerError(null)
        setAuthServerError(null)
    }, [location])

    let contextData = {
        login: login,
        loginEmail: loginEmail,
        setLoginEmail: setLoginEmail,
        loginPassword: loginPassword,
        setLoginPassword: setLoginPassword,
        loginError: loginError,
        authTokens: authTokens, 
        logout: logout,
        authServerError: authServerError,
        unexpectedLogoutError: unexpectedLogoutError,
        verifyEmail: verifyEmail,
        initialVerificationError: initialVerificationError,
        initialVerificationServerError: initialVerificationServerError,
        initialVerificationSuccessful: initialVerificationSuccessful,
        setAuthSignUpServerError: setAuthSignUpServerError,
        authSignUpServerError: authSignUpServerError,
        BACKEND_DOMAIN: BACKEND_DOMAIN,
        setLoginError: setLoginError,
        userId: userId,
        updateToken: updateToken,
        loginButtonDisabled: loginButtonDisabled,
    }

    return(
        <AuthContext.Provider value={contextData}>
            {children}
        </AuthContext.Provider>
    )}