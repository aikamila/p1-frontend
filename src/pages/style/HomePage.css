:root {
  --nice-blue: rgb(54, 155, 250);
  --background-gray: rgb(250, 250, 250);
  --disabled-blue: rgb(157, 200, 240);
  --fonts: "Raleway", "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell",
    "Fira Sans", "Droid Sans", "Helvetica Neue", sans-serif;
}

.home-page__email_verified_notification {
  margin: 0%;
  text-align: center;
  padding-bottom: 4px;
  padding-top: 3px;
  background-color: var(--background-gray);
}
