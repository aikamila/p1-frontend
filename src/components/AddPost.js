import React, { useState, useContext } from 'react'
import { Link } from 'react-router-dom'
import api from '../api/PostsComments'
import AuthContext from '../context/AuthContext'
import { FiLoader } from "react-icons/fi"
import './style/AddEditPost.css'
import HomeHeader from './HomeHeader'


const AddPost = () => {
  const {userId, authTokens} = useContext(AuthContext)
  const [success, setSuccess] = useState(false)
  const [failure, setFailure] = useState("")
  const [disabled, setDisabled] = useState(true)
  const [loading, setLoading] = useState(false)
  const [text, setText] = useState("")
  const [textTooLong, setTextTooLong] = useState(false)
  const [minCharsCounter, setMinCharsCounter] = useState(0)
  const countInput = (e) => {
    setFailure("")
    const result = 5000 - e.target.value.length
    if(result>=0){
      setText(e.target.value)
      setTextTooLong(false)
    }else{
      setTextTooLong(true)
    }
    if(e.target.value.trim().length < 30){
      setDisabled(true)
      setMinCharsCounter(e.target.value.trim().length)
    }else{
      setDisabled(false)
      setMinCharsCounter(30)
    }
  }
  const addPost = async() => {
    setTextTooLong(false)
    setDisabled(true)
    setLoading(true)
    try{
      const response = await api.post(`/`, {text: text}, {
        headers: {
          Authorization: `Bearer ${authTokens.access}`
        }
      })
      if(response.status !== 201){
        throw Error("Something went wrong")
      }
      setSuccess(true)
    }catch(err){
      if(err.response && err.response.status===400){
        setFailure("Hmm... Your post seems to be invalid")
      }
      else{
        setFailure("Something went wrong. Try to submit the post later.")
      }
    }
    setLoading(false)
    setDisabled(false)
  }
  return (
    <>
      <HomeHeader/>
      <main className='post-add__background'>
        <div className='post-add__center'>
          {success ? 
          <div className='post-add__success'>
            <p>Your post was created successfully</p>
            <Link className="post-add__success_link" to="/home">Go back to the homepage</Link>
            <Link className="post-add__success_link" to={`/home/account/${userId}`}>See your posts</Link>
          </div>:
          <>
            <h1 className='post-add__title'>Add a public post</h1>
            <p className='post-add__description' id="post-description">Remember to include a full description of your offer and your requirements.</p>
            <div className='post-add__content_input_wrapper'>
              <textarea aria-label='your post' aria-describedby='post-description' placeholder='Share your idea with the world...' 
              value={text} onChange={(e) => (countInput(e))} className='post-add__content_input'></textarea>
            </div>
            {minCharsCounter !== 30 && <><span className='post-add__input_counter'>{minCharsCounter}/30</span><br/></>}
            {textTooLong && <p role="alert" className='post-add__input_err'>Your post is too long</p>} 
            {failure && <p role="alert" className='post-add__input_err'>{failure}</p>}
            <button disabled={disabled} onClick={addPost} aria-label="Add your post" className='post-add__add_button'>Add</button>
            <div className='post-add__loading_sign_div'>
              {loading && <FiLoader role="img" aria-label="loading" className='post-add__loading_sign --spinner'></FiLoader>}
            </div>
          </>
          }
        </div>
      </main>
    </>
  )
}

export default AddPost