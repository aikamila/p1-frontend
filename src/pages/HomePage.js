import React,{useContext} from 'react'
import AuthContext from '../context/AuthContext'
import { Switch, Route } from 'react-router-dom'
import ListPosts from '../components/ListPosts'
import PostDetails from '../components/PostDetails'
import UpdatePost from '../components/UpdatePost'
import AddPost from '../components/AddPost'
import UserAccount from '../components/UserAccount'
import './style/HomePage.css'


const HomePage = () => {
    let {initialVerificationSuccessful} = useContext(AuthContext)

    return (
        <>
            {initialVerificationSuccessful && <p role="alert">Your email was successfully verified!</p>}
            <Switch>
                <Route path="/home/account/:id" component={UserAccount} />
                <Route path="/home/post/:postId/update" component={UpdatePost}/>
                <Route path="/home/post/add" component={AddPost} />
                <Route path="/home/post/:postId" component={PostDetails}/>
                <Route path="/home">
                    <ListPosts />
                </Route>
            </Switch>
        </>
    )
}

export default HomePage

