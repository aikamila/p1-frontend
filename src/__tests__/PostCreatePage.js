import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import userEvent from '@testing-library/user-event'
import AuthContext from '../context/AuthContext'
import HomePage from '../pages/HomePage'
import { server } from '../mocks/server'
import { rest } from 'msw'
import renderer from 'react-test-renderer'
import AddPost from '../components/AddPost'

const renderAddPostPage = () => {
    const history = createMemoryHistory();
    history.push("/home/post/add")
    const authTokens = {refresh: 'valid-refresh', access: 'valid-access'}
    return render(
    <Router history={history}>
        <AuthContext.Provider value={{logout: null, initialVerificationSuccessful: false, authTokens: authTokens,
        userId: 1, updateToken: null}}>
            <HomePage />
        </AuthContext.Provider>
    </Router>)
}

test("all elements are initially rendered", () => {
    renderAddPostPage()
    expect(screen.getByRole("textbox", {name:/your post/i})).toBeInTheDocument()
    expect(screen.getByPlaceholderText(/share your idea with the world/i)).toBeInTheDocument()
    expect(screen.getByRole("button", {name: /post/i})).toBeInTheDocument()
    expect(screen.getByRole("button", {name: /post/i})).toBeDisabled()
    expect(screen.getByText("0/30")).toBeInTheDocument()
})

test("submit button gets enabled after the user types minimmum 30 characters (without whitespaces at the end and at the beginning)", () => {
    // testing the minimum word counter at the same time
    renderAddPostPage()
    const textbox = screen.getByRole("textbox")
    userEvent.type(textbox, " ")
    expect(screen.getByRole("button", {name: /post/i})).toBeDisabled()
    expect(screen.getByText("0/30")).toBeInTheDocument()
    userEvent.type(textbox, " ".repeat(28))
    expect(screen.getByRole("button", {name: /post/i})).toBeDisabled()
    expect(screen.queryByText("0/30")).toBeInTheDocument()
    userEvent.type(textbox, "hi whats going on everybody  ")
    expect(screen.getByRole("button", {name: /post/i})).toBeDisabled()
    expect(screen.queryByText("27/30")).toBeInTheDocument()
    userEvent.type(textbox, "?")
    expect(screen.getByRole("button", {name: /post/i})).not.toBeDisabled()
    expect(screen.queryByText("/30")).not.toBeInTheDocument()
})

test("maximum lenght of input is 5000", () => {
    renderAddPostPage()
    const textbox = screen.getByRole("textbox")
    fireEvent.change(textbox, {target: {value: "a".repeat(4999)}})
    expect(screen.queryByText(/post is too long/i)).not.toBeInTheDocument()
    userEvent.type(textbox, "a")
    expect(screen.queryByText(/post is too long/i)).not.toBeInTheDocument()
    expect(textbox).toHaveValue("a".repeat(5000))
    userEvent.type(textbox, "a")
    expect(screen.queryByText(/post is too long/i)).toBeInTheDocument()
    expect(screen.getByRole("alert")).toBeInTheDocument()
    expect(textbox).toHaveValue("a".repeat(5000))
    userEvent.type(textbox, "{backspace}")
    expect(screen.queryByText(/post is too long/i)).not.toBeInTheDocument()
    expect(screen.queryByRole("alert")).not.toBeInTheDocument()
    expect(textbox).toHaveValue("a".repeat(4999))
})

test("post successfully created", async () => {
    renderAddPostPage()
    const textbox = screen.getByRole("textbox")
    userEvent.type(textbox, "My new post my new post my new post my new post")
    userEvent.click(screen.getByRole("button", {name: /post/i}))
    expect(screen.getByRole("button", {name: /post/i})).toBeDisabled()
    expect(screen.getByRole("img", {name: /loading/i})).toBeInTheDocument()
    expect(await screen.findByText(/your post was created successfully/i)).toBeInTheDocument()
    expect(screen.getByRole("link", {name: /to the homepage/i})).toHaveAttribute("href", "/home")
    expect(screen.getByRole("link", {name: /your posts/i})).toHaveAttribute("href", "/home/account/1")
    expect(screen.queryByRole("img", {name: /loading/i})).not.toBeInTheDocument()
    expect(screen.queryByRole("textbox")).not.toBeInTheDocument()
    expect(screen.queryByRole("button", {name: /post/i})).not.toBeInTheDocument()
})

test("post isn't created because server responds wth 400", async () => {
    renderAddPostPage()
    const textbox = screen.getByRole("textbox")
    userEvent.type(textbox, "Some invalid post some invalid post some invalid post")
    userEvent.click(screen.getByRole("button", {name: /post/i}))
    expect(screen.getByRole("button", {name: /post/i})).toBeDisabled()
    expect(screen.getByRole("img", {name: /loading/i})).toBeInTheDocument()
    expect(await screen.findByText(/seems to be invalid/i)).toBeInTheDocument()
    expect(screen.getByRole("alert")).toBeInTheDocument()
    expect(screen.queryByRole("link", {name: / go to the homepage/i})).not.toBeInTheDocument()
    expect(screen.queryByRole("link", {name: /your posts/i})).not.toBeInTheDocument()
    expect(screen.queryByRole("img", {name: /loading/i})).not.toBeInTheDocument()
    expect(screen.queryByRole("textbox")).toBeInTheDocument()
    expect(screen.queryByRole("textbox")).toHaveValue("Some invalid post some invalid post some invalid post")
    expect(screen.queryByRole("button", {name: /post/i})).toBeInTheDocument()
    expect(screen.queryByRole("button", {name: /post/i})).not.toBeDisabled()
    userEvent.type(textbox, "a")
    expect(screen.queryByText(/seems to be invalid/i)).not.toBeInTheDocument()
    expect(screen.queryByRole("alert")).not.toBeInTheDocument()
})

test("post isn't created beacuse server sends an invalid response", async () => {
    server.use(
        rest.post('https://arcane-spire-03245.herokuapp.com/api/services/posts/', (req, res, ctx) => {
          return res.once(
              ctx.status(401)
          )
        }),
    )
    renderAddPostPage()
    const textbox = screen.getByRole("textbox")
    userEvent.type(textbox, "My new post my new post my new post my new post")
    userEvent.click(screen.getByRole("button", {name: /post/i}))
    expect(screen.getByRole("button", {name: /post/i})).toBeDisabled()
    expect(screen.getByRole("img", {name: /loading/i})).toBeInTheDocument()
    expect(await screen.findByText(/try to submit the post later/i)).toBeInTheDocument()
    expect(screen.getByRole("alert")).toBeInTheDocument()
    expect(screen.queryByRole("link", {name: / go to the homepage/i})).not.toBeInTheDocument()
    expect(screen.queryByRole("link", {name: /your posts/i})).not.toBeInTheDocument()
    expect(screen.queryByRole("img", {name: /loading/i})).not.toBeInTheDocument()
    expect(screen.queryByRole("textbox")).toBeInTheDocument()
    expect(screen.queryByRole("textbox")).toHaveValue("My new post my new post my new post my new post")
    expect(screen.queryByRole("button", {name: /post/i})).toBeInTheDocument()
    expect(screen.queryByRole("button", {name: /post/i})).not.toBeDisabled()
    userEvent.type(textbox, "a")
    expect(screen.queryByText(/try to submit the post later/i)).not.toBeInTheDocument()
    expect(screen.queryByRole("alert")).not.toBeInTheDocument()
})

test("alert about too many characters disappears when the user submits their post", async () => {
    // asserts that the user never sees 2 errors at the same time (they can get another error
    // while submitting the post)
    server.use(
        rest.post('https://arcane-spire-03245.herokuapp.com/api/services/posts/', (req, res, ctx) => {
          return res.once(
              ctx.status(500)
          )
        }),
    )
    renderAddPostPage()
    const textbox = screen.getByRole("textbox")
    fireEvent.change(textbox, {target: {value: "a".repeat(4999)}})
    userEvent.type(textbox, "aa")
    expect(screen.queryByText(/post is too long/i)).toBeInTheDocument()
    userEvent.click(screen.getByRole("button", {name: /post/i}))
    expect(screen.queryByText(/post is too long/i)).not.toBeInTheDocument()
    expect(await screen.findByText(/try to submit the post later/i)).toBeInTheDocument()
    expect(screen.getAllByRole('alert')).toHaveLength(1)
})

test("add post form is rendered correctly", () => {
    const history = createMemoryHistory();
    history.push("/auth")
    const page = renderer
      .create(
        <Router history = {history}>
          <AuthContext.Provider value={{userId: 3, authTokens: {access: 'valid-token', refresh: 'refresh-token'}}}>
            <AddPost />
          </AuthContext.Provider>
        </Router>
      )
      .toJSON()
      expect(page).toMatchSnapshot()
  })