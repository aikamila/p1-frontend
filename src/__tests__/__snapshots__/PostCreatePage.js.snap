// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add post form is rendered correctly 1`] = `
Array [
  <header
    className="home-header__background"
    role="banner"
  >
    <div>
      <img
        alt="shangify brand logo"
        className="logo__img1"
        src="shangify_logo.png"
      />
      <img
        alt=""
        className="logo__img2"
        src="shangify_brand_name.png"
      />
    </div>
    <nav
      aria-label="main navigation"
      className="home-header__nav"
      role="navigation"
    >
      <a
        aria-label="add a post"
        href="/home/post/add"
        onClick={[Function]}
        style={
          Object {
            "pointerEvents": "auto",
          }
        }
        title="Add a post"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4z"
          />
          <path
            d="M12 2C6.486 2 2 6.486 2 12s4.486 10 10 10 10-4.486 10-10S17.514 2 12 2zm0 18c-4.411 0-8-3.589-8-8s3.589-8 8-8 8 3.589 8 8-3.589 8-8 8z"
          />
        </svg>
      </a>
      <a
        aria-label="homepage"
        href="/home"
        onClick={[Function]}
        style={
          Object {
            "pointerEvents": "auto",
          }
        }
        title="Homepage"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 22h14a2 2 0 0 0 2-2v-9a1 1 0 0 0-.29-.71l-8-8a1 1 0 0 0-1.41 0l-8 8A1 1 0 0 0 3 11v9a2 2 0 0 0 2 2zm5-2v-5h4v5zm-5-8.59 7-7 7 7V20h-3v-5a2 2 0 0 0-2-2h-4a2 2 0 0 0-2 2v5H5z"
          />
        </svg>
      </a>
      <a
        aria-label="my account"
        href="/home/account/3"
        onClick={[Function]}
        style={
          Object {
            "pointerEvents": "auto",
          }
        }
        title="My account"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 14s-1 0-1-1 1-4 6-4 6 3 6 4-1 1-1 1H3zm5-6a3 3 0 1 0 0-6 3 3 0 0 0 0 6z"
          />
        </svg>
      </a>
      <button
        disabled={false}
      >
        Log out
      </button>
    </nav>
  </header>,
  <main
    className="post-add__background"
    role="main"
  >
    <div
      className="post-add__center"
    >
      <h1
        className="post-add__title"
      >
        Add a public post
      </h1>
      <p
        className="post-add__description"
        id="post-description"
      >
        Remember to include a full description of your offer and your requirements.
      </p>
      <div
        className="post-add__content_input_wrapper"
      >
        <textarea
          aria-describedby="post-description"
          aria-label="your post"
          className="post-add__content_input"
          onChange={[Function]}
          placeholder="Share your idea with the world..."
          value=""
        />
      </div>
      <span
        className="post-add__input_counter"
      >
        0
        /30
      </span>
      <br />
      
      <button
        aria-label="Add your post"
        className="post-add__add_button"
        disabled={true}
        onClick={[Function]}
      >
        Add
      </button>
      <div
        className="post-add__loading_sign_div"
      />
    </div>
  </main>,
  <footer
    className="footer__background"
    role="contentinfo"
  >
    <div
      className="footer__container"
    >
      <span>
        Â©Copyright 
        2022
      </span>
    </div>
  </footer>,
]
`;
